lib/Sema/CSDiagnostics.h

```diff
@@ -417,9 +417,10 @@ class MissingConformanceFailure final : public RequirementFailure {
 public:
   MissingConformanceFailure(Expr *expr, ConstraintSystem &cs,
                             ConstraintLocator *locator,
-                            std::pair<Type, Type> conformance)
-      : RequirementFailure(cs, expr, RequirementKind::Conformance,
-                           conformance.first, conformance.second, locator) {}
+                            std::pair<Type, Type> conformance,
+                            RequirementKind kind)
```

Comment:
In `RequirementFailure` constructor, we pass in the `RequirementKind` and check:
```cpp
    auto reqElt = locator->castLastElementTo<LocatorPathElt::AnyRequirement>();
    assert(reqElt.getRequirementKind() == kind);
```
If we're going to get `kind` from the locator anyway, is this necessary? `kind` isn't used anywhere else in the constructor. I think it makes more sense to perform assertions when constructing the various subclasses of `RequirementFailure`. What do you think?

Reply:
Yes, that's why I mentioned that locator has `kind` now, I think we can just use that instead of passing it as an argument. Originally `TypeParameterRequirement` didn't have any information, but once I have extended it first with index then kind I didn't update all of the places :/